mosaic/mosaic.go

```diff
@@ -0,0 +1,423 @@
+
+			// Append to output.
+			output.WriteString(
+				ansi.Style{}.ForegroundColor(block.BestFgColor).BackgroundColor(block.BestBgColor).Styled(string(block.BestSymbol)),
+			)
+		}
+		output.WriteString("\n")
+	}
+
+	return output.String()
+}
+
+// createPixelBlock extracts a 2x2 block of pixels from the image.
+func (m *Mosaic) createPixelBlock(img image.Image, x, y int) *pixelBlock {
+	block := &pixelBlock{}
+
+	// Extract the 2x2 pixel grid.
+	for dy := 0; dy < 2; dy++ {
+		for dx := 0; dx < 2; dx++ {
+			block.Pixels[dy][dx] = m.getPixelSafe(img, x+dx, y+dy)
+		}
+	}
+
+	return block
+}
+
+// findBestRepresentation finds the best block character and colors for a 2x2 pixel block.
+func (m *Mosaic) findBestRepresentation(block *pixelBlock, availableBlocks []Block) {
+	// Simple case: use only foreground/background colors.
+	if m.useFgBgOnly {
+		// Just use the upper half block with top pixels as background and bottom as foreground.
+		block.BestSymbol = '▀'
+		block.BestBgColor = m.averageColors([]color.Color{block.Pixels[0][0], block.Pixels[0][1]})
+		block.BestFgColor = m.averageColors([]color.Color{block.Pixels[1][0], block.Pixels[1][1]})
+		return
+	}
+
+	// Determine which pixels are "set" based on threshold.
+	pixelMask := [2][2]bool{}
+	for y := 0; y < 2; y++ {
+		for x := 0; x < 2; x++ {
+			// Calculate luminance.
+			luma := rgbaToLuminance(block.Pixels[y][x])
+			pixelMask[y][x] = luma >= m.thresholdLevel
+		}
+	}
+
+	// Find the best matching block character.
+	bestChar := ' '
+	bestScore := math.MaxFloat64
+
+	for _, blockChar := range availableBlocks {
+		score := 0.0
+		for i := 0; i < 4; i++ {
+			y, x := i/2, i%2
+			if blockChar.Coverage[i] != pixelMask[y][x] {
+				score += 1.0
+			}
+		}
+
+		if score < bestScore {
+			bestScore = score
+			bestChar = blockChar.Char
+		}
+	}
+
+	// Determine foreground and background colors based on the best character.
+	var fgPixels, bgPixels []color.Color
+
+	// Get the coverage pattern for the selected character.
+	var coverage [4]bool
+	for _, b := range availableBlocks {
+		if b.Char == bestChar {
+			coverage = b.Coverage
+			break
+		}
+	}
+
+	// Assign pixels to foreground or background based on the character's coverage.
+	for i := 0; i < 4; i++ {
+		y, x := i/2, i%2
+		if coverage[i] {
+			fgPixels = append(fgPixels, block.Pixels[y][x])
+		} else {
+			bgPixels = append(bgPixels, block.Pixels[y][x])
+		}
+	}
+
+	// Calculate average colors.
+	if len(fgPixels) > 0 {
+		block.BestFgColor = m.averageColors(fgPixels)
+	} else {
+		// Default to black if no foreground pixels.
+		block.BestFgColor = color.Black
+	}
+
+	if len(bgPixels) > 0 {
+		block.BestBgColor = m.averageColors(bgPixels)
+	} else {
+		// Default to black if no background pixels.
+		block.BestBgColor = color.Black
+	}
+
+	block.BestSymbol = bestChar
+}
+
+// averageColors calculates the average color from a slice of colors.
+func (m *Mosaic) averageColors(colors []color.Color) color.Color {
```

Comment:
can make this 

```suggestion
func (m *Mosaic) averageColors(colors ...color.Color) color.Color {
```
 
so you don't need to create a slice in line 263 for example